clang/test/CodeGenSYCL/registered-kernel-names.cpp

```diff
@@ -0,0 +1,71 @@
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -triple spir64 -o - %s | FileCheck %s
+
+// This test checks if the sycl_registered_kernels named metadata and
+// associated entries are generated for registered kernel names.
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void foo() {
+}
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-nd-range-kernel", 0)]]
+void ff_4() {
+}
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void iota(int, int *) {
+}
+
+template <typename T>
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void tempfoo(T pt);
+
+template <typename T>
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void tempfoo2(T pt) {
+  T t;
+}
+
+template void tempfoo2<int>(int);
+
+template <>
+void tempfoo2(float f);
+
+template <>
+void tempfoo2(short) { }
+
+template <int N>
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void tempfoo3() {
+  (void)N;
+}
+
+namespace N {
+[[__sycl_detail__::__registered_kernels__(
+  {"foo", foo},
+  {"iota", (void(*)(int, int *))iota},
+  {"decl temp", tempfoo<int>},
+  {"inst temp", tempfoo2<int>},
+  {"decl spec", tempfoo2<float>},
+  {"def spec", tempfoo2<short>},
+  {"foo3", ff_4},
+  {"nontype", tempfoo3<5>}
+)]];
+}
+
+// Check for the presence of sycl_registered_kernels named metadata.
```

Comment:
Can we also check that kernel body exists in the module for a free function that is not instantiated and only mentioned by the `__registered_kernels__` attribute?

Reply:
Done.   Added ``bar`` in the test.